docs/specs/workflow-customization.md

```diff
@@ -0,0 +1,1032 @@
+- We want to keep the guarantee that the existing configuration ensures every new user has some certain identities and authenticators.
+- We want to be able to fulfill the authentication flows in existing consumer apps
+
+### What is a signup flow
+- Create 1 or more Identities. Later on, the User identify themselves with one of the Identities.
+- Create 0 or more Authenticators. The User authenticates themselves with one of the Authenticators if needed.
+- (Optional) Collect user profile (i.e. standard attributes and custom attributes)
+
+### What is a login flow
+- Identify the User with an Identity
+- Depending on the Identity, authenticator the User with the Authenticators. Note that the pre-selected Authenticator is usually associated with the Identity.
+- (Optional) Further authenticate the User with **other** Authenticators.
+
+Suppose the User has a Email Login ID Identity `johndoe@gmail.com`, a Email OOB-OTP Authenticator `johndoe@gmail.com`, a Phone Login ID Identity `+85298765432`, a Phone OOB-OTP Authenticator `+85298765432`, a Password Authenticator, and a OAuth Identity `johndoe@gmail.com`.
+
+If the User identifies themselves with the Email Login ID Identity `johndoe@gmail.com`, then the User can authenticate themselves with:
+1. Email OOB-OTP Authenticator `johndoe@gmail.com`
+2. Password Authenticator
+3. Phone OOB-OTP Authenticator `+85298765432`. Note that this Authenticator is NOT associated with the identifying Identity, but it can also be used to authenticate the User.
+
+If the User identifies themselves with the OAuth Identity `johndoe@gmail.com`, then the User DOES NOT need to authenticate themselves. This is how most other applications work.
+
+### What is a reauth flow
+- Authenticate the User with any Authenticators.
+
+### What is account linking
+
+Account linking happens in a login flow.
+Each login flow can optionally specify the conditions when account linking can occur.
+If no conditions are specified or no conditions are matched, an error is returned, telling the end-user to sign in with the existing identity instead.
+
+The login flow is then proceeded as if the existing identity were selected.
+
+At the end of the flow, the new identity is added to the account.
+
+### Design of the configuration
+
+#### Design overview
+
+- A flow has one or more `steps`.
+- A step MAY optionally have an `id`.
+- A step must have a `type`.
+- A `type` step is specific to the kind of the flow. For example, only SignupFlow has the `type: user_profile` step.
+- Some steps allow branching. Those steps have `one_of`.
+- The branch of a step MAY optionally have zero or more `steps`.
+
+#### SignupFlow
+
+Example:
+
+```yaml
+signup_flows:
+- id: default_signup_flow
+  steps:
+  - id: setup_identity
+    type: identify
+    one_of:
+    - identification: phone
+      steps:
+      - type: authenticate
+        one_of:
+        - authentication: primary_oob_otp_sms
+          target_step: setup_identity
+      - step: verify
```

Comment:
@louischan-oursky The verify step here is missing `type`.

Reply:
Fixed.